### PR TITLE
Pc 363 postcode validation and refactor

### PR DESCRIPTION
**Pull request by xnw9**
- Include postcode filtering from Chiara's branch
- Add validation (no sanitation required)
- Validation tests to be added?

### CODE REVIEW

File: help_to_heat/frontdoor/interface.py [hunk 1]
```diff
@@ -71,6 +77,47 @@ class SuccessSchema(marshmallow.Schema):
...
+    if (offset):
+        url += f"&offset={offset}"
+    return url
+
+def get_addresses_from_api(postcode) -> List[Dict]:
+    try:
+        max_results_number = 100
+        offset = 0
+
+        url = get_url(postcode=postcode, offset=offset, max_results=max_results_number)
```
Comment (JoshAdamPowell) [inline help_to_heat/frontdoor/interface.py:98]: not really an issue but no need for named args here

File: help_to_heat/frontdoor/interface.py [hunk 2]
```diff
@@ -71,6 +77,47 @@ class SuccessSchema(marshmallow.Schema):
...
+        response = requests.get(url)
+        response.raise_for_status()
+        json_response = response.json()
+
+        total_results_number = json_response['header']["totalresults"]
+        api_results = json_response['results']
+        api_results_all = api_results
+
+        if total_results_number > max_results_number:
```
Comment (JoshAdamPowell) [inline help_to_heat/frontdoor/interface.py:107]: maybe it's Monday morning brain but I'm finding this a little tricky to follow, can we put it in a function so we can describe what it's doing?

File: help_to_heat/frontdoor/interface.py [hunk 3]
```diff
@@ -108,23 +155,194 @@ def create_referral(self, session_id):
...
+             "2720": "2745",
+             "2725": "2745",
+             "2730": "2745",
+             "2735": "2745",
+             "3305": "3300", 
+             "3310": "3300",
+             "3325": "3300",  
+             "3330": "3300", 
+             }
+        # Map:
```
Comment (JoshAdamPowell) [inline help_to_heat/frontdoor/interface.py:364]: Is this copied from somewhere?If so, can we have a link to it in a comment? It's all a bit bewildering!

**Review by JoshAdamPowell — CHANGES_REQUESTED** [review]

File: help_to_heat/frontdoor/interface.py (hunk 3 above)
Comment (LadunOmideyiSoftwire) [reply help_to_heat/frontdoor/interface.py:364]: Yes, it's hardcoded Local Authority mergers. The old codes are just mapped to the new codes.

File: help_to_heat/frontdoor/interface.py [hunk 4]
```diff
@@ -71,6 +77,47 @@ class SuccessSchema(marshmallow.Schema):
     success = marshmallow.fields.Boolean()
 
 
+def get_url(postcode: str, offset: int, max_results: int) -> str:
```
[the remark this answers is not in the document]
Comment (xnw9) [reply help_to_heat/frontdoor/interface.py:85]: Makes sense! I've removed all the types :)

File: help_to_heat/frontdoor/interface.py (hunk 3 above)
Comment (xnw9) [reply help_to_heat/frontdoor/interface.py:364]: A comment is added to explain :)

File: help_to_heat/frontdoor/interface.py (hunk 2 above)
Comment (xnw9) [reply help_to_heat/frontdoor/interface.py]: Agree it's a bit hard to follow.. how about adding a comment? I tried pulling out a function but that didn't look good :( I tried refactoring the `get_url` function to `get_json_response_from_url`, and it looks slightly better?

Comment (JoshAdamPowell) [reply help_to_heat/frontdoor/interface.py]: Comment is acceptable instead of pulling out a function :)

Comment (xnw9) [reply help_to_heat/frontdoor/interface.py]: Cool, added some explanation and kept the latest refactoring :)

**Review by JoshAdamPowell — APPROVED** [review]
